Eldev support suggestion (draft)

ikappaki opened the pull request:
Hi,

would you consider using a build tool for `moldable-emacs` that will greatly facilitate development any further?

I found it difficult earlier to run the tests while looking at #14, and though it will be a good idea to introduce [Eldev](https://github.com/doublep/eldev) for the task, if not for everything else.

The only changes required are to add the `Eldev` configuration file and include the package info at the top of `moldable-emacs.el`.

Then the tests can be run with `eldev test`, with the added benefit they are now run in isolation, only using the dependent packages and nothing else in a clean setup. The tests are further consider by the CI as a GitHub action with `.github/workflows/test.yml' so as to run every time there is a commit.

Moreover, `Eldev` has more to offer such as package creation, linting etc, which greatly helps with the whole management lifecycle.

Let me know what you think. `Eldev` is a quite mature build tool and is used by quite popular packages such as [CIDER](https://github.com/clojure-emacs/cider).

Thanks,

https://github.com/doublep/eldev

ag91 commented:
Oh, thanks! I think it is a wonderful idea. I have been playing around so far, but I guess it would improve life for everybody if we have some easy way to test things.
Let me have a look in eldev and try it out.

ag91 commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,34 @@
+
+    steps:
+    - name: Set up Emacs
+      uses: purcell/setup-emacs@master
```
I am unfamiliar with github workflows: this picks purcell's Github action to setup an Emacs version for testing.
Does it mean that if purcell deletes his repo this will break?

ikappaki replied:
Yes, but this would be a most unprecedented event. Nevertheless, if it does we can always switch to an alternative Emacs installation (there a few other choices around) or a fork of the removed repo.

Just to elaborate a little bit on how it works, this GitHub workflow sets up Continuous Integration (CI) on the GitHub servers. When a git push or pull event happens, GitHub allocates a clean Ubuntu server to run some *actions*, in our case we checkout the codebase, install Emacs using Purcell's action, and run `eldev test` on it. If the tests succeeds, then we get a green icon on GitHub when browsing the repository, otherwise a red cross indicating the latest push or pull request has broken something. The submitter then has to go back and fix the break. Thus, the integrity of the codebase is maintained across commits.

Back to our point, there are quite a few actions available on GitHub created by individuals and are used by thousands of GitHub projects, If the author deletes such a repository, then thousands of projects will fail CI. Although this is a possible scenario,it is most unlikely that someone will delete an action many depend on it without a good reason. These actions are really simple scripts that can be reproduce anytime or replaced by a fork someone has available.

Please also note this workflow sits on top of `Eldev`, we don't need to set up Continuous Integration. We can still use `eldev test` on our development machine, I thought it would have been nice to set up CI anyway. now that we have `Eldev` configured in the codebase :)

ag91 replied:
Thanks for the detailed explanation! I was just curious to see if I understood things right. It all sounds great, thanks for introducing me to Eldev!!

ikappaki commented:
> Oh, thanks! I think it is a wonderful idea. I have been playing around so far, but I guess it would improve life for everybody if we have some easy way to test things. Let me have a look in eldev and try it out.

Please also try out the linter with `eldev lint` and `eldev package` to create a package out of the project (that would be very useful when the project is ready to publish at MELPA).

I can have a look at the linter errors if you are happy to go ahead with `Eldev` :)

Thanks

ag91 commented:
Thanks, I tried all the commands you suggested and it looks amazing. I think we could merge as is and then I can fix the package details. I was planning to give a look at the linters complaining at some point but I didn't allocate the time yet :P 
For now I am giving a look at the (long) list to make sure I don't introduce more XD

Thanks again! You contribution is making this extension much more mature :D

ikappaki commented:
> Thanks, I tried all the commands you suggested and it looks amazing. I think we could merge as is and then I can fix the package details. I was planning to give a look at the linters complaining at some point but I didn't allocate the time yet :P For now I am giving a look at the (long) list to make sure I don't introduce more XD
> 
> Thanks again! You contribution is making this extension much more mature :D

Glad you liked it :) I've made some minor changes (removed tree-sitter as main dependencies, but marked them as required in the Eldev file for the purpose of running the tests. I've also dropped a line in the readme about running the tests. Squashed and ready to commit when you feel like it :)
